tools/clang/lib/CodeGen/CGExprScalar.cpp

```diff
@@ -3035,7 +3035,7 @@ Value *ScalarExprEmitter::EmitShl(const BinOpInfo &Ops) {
                       Ops.Ty->hasSignedIntegerRepresentation();
   bool SanitizeExponent = CGF.SanOpts.has(SanitizerKind::ShiftExponent);
   // OpenCL 6.3j: shift values are effectively % word size of LHS.
-  if (CGF.getLangOpts().OpenCL)
+  if (CGF.getLangOpts().OpenCL || CGF.getLangOpts().HLSL) // HLSL Change
     RHS =
         Builder.CreateAnd(RHS, GetWidthMinusOneValue(Ops.LHS, RHS), "shl.mask");
   else if ((SanitizeBase || SanitizeExponent) &&
```

Comment:
I find myself thinking the CL behavior exactly matching what we need for HLSL seems too good to be true. I'm not familiar with these sanitizer settings. Where do they get set? How do the checks affect the codegen?

Reply:
That is the undefined behavior sanitizer instrumentation, which isn't supported by HLSL (yet!).

OpenCL's definition for bit shifts matches HLSL's probably more because it matches common hardware implementations.